Make Makefiles in `test/images/` compatible with multiple architectures

mkumatag opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#pull-request-process and developer guide https://github.com/kubernetes/community/blob/master/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/community/blob/master/contributors/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**:
This PR is for making test images multi architecture for different platforms like amd64, arm, arm64, ppc64le
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes  #31331

**Special notes for your reviewer**:
- Actual tests need to be modified to use these images based on the architecture later.
- Not covering the cross building of docker images for `s390x` platform due to problem faced while running containers with `qemu-s390x-static`
- Will submit separate PR for `volume and pet` test images
- This PR depends on - https://github.com/kubernetes/ingress/pull/587

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```NONE```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44910)
<!-- Reviewable:end -->

mkumatag commented:
/cc @luxas @ixdy

luxas reviewed (changes requested):
Thanks for you awesome work!
I didn't get to review this line-by-line, but some overall comments:
 - Let's focus on `test/images` in this PR, submit `examples/` changes in an other PR
 - The amount of duplicated code currently is kind of insane. Please combine all test images to one Makefile only that may delegate tasks if needed to a sub-Makefile per image. The version/image name information should also exist in the per-image sub-Makefile.

cc @gajju26 on s390x support.

mkumatag commented:
> Thanks for you awesome work!
> I didn't get to review this line-by-line, but some overall comments:
> 
> - Let's focus on test/images in this PR, submit examples/ changes in an other PR

The reason behind for this is - I was targeting all the images needed for conformance tests. I can submit another PR if that is the concern here!
> - The amount of duplicated code currently is kind of insane. Please combine all test images to one Makefile only that may delegate tasks if needed to a sub-Makefile per image. The version/image name information should also exist in the per-image sub-Makefile.

Not really sure how far this is feasible because we have deviation for building container images because some need golang code build + dockerfile and some are just plain dockerfile.

I started exploring this option but was stuck while importing varibles from sub-Makefile to parent Makefile. AFAIK its not possible to read that way, let me know if you have explored anything with that.
> cc @gajju26 on s390x supp

luxas commented:
> The reason behind for this is - I was targeting all the images needed for conformance tests. I can submit another PR if that is the concern here!

Yes, please do examples in another PR

> Not really sure how far this is feasible because we have deviation for building container images because some need golang code build + dockerfile and some are just plain dockerfile.

What about matching the image name with the directory name and putting version in a file called `VERSION` and having specialized subtasks in makefiles or scripts to call before building?

I think we should come up with some way to minimize the duplication here
cc @ixdy

mkumatag commented:
I spent some more time and have a idea to reduce the code duplication.

File structure looks like this:

```
[manjunath@kumatagi test]$ ls -R
.:
build-image.sh  Makefile  test-image1  test-image2  test-image3  test-image4

./test-image1:
BASEIMAGE  bootstrap.sh  NAME  VERSION

./test-image2:
......
.......
```

Makefile:

```
[manjunath@kumatagi test]$ cat Makefile 
REGISTRY ?= gcr.io/google_containers
GOARM=7
GOLANG_VERSION=1.7.5
export

SUBDIRS := $(subst /NAME,,$(shell find . -name 'NAME'))

all: $(SUBDIRS)
$(SUBDIRS):
	$(MAKE) build-image TEST=$@

build-image:
	./build-image.sh container $(TEST)

.PHONY: all $(SUBDIRS)
[manjunath@kumatagi test]$ 

```

and here the Makefile is responsible for building all the images from subfolders like test-image1, test-image2 etc..

To avoid any dup makefiles in under all the test images, prefer creating files like BASEIMAGE, NAME, VERSION, bootstrap.sh..

```
[manjunath@kumatagi test-image1]$ more *
::::::::::::::
BASEIMAGE
::::::::::::::
amd64=busybox
arm=armhf/busybox
arm64=aarch64/busybox
ppc64le=ppc64le/busybox
s390x=s390x/busybox
::::::::::::::
bootstrap.sh
::::::::::::::
::::::::::::::
NAME
::::::::::::::
image1
::::::::::::::
VERSION
::::::::::::::
0.1
[manjunath@kumatagi test-image1]$ 
```

`bootstrap.sh` script is responsible for dealing with any prereq required before running `docker build` in that directory like building go code, downloading qemu etc..

Let me know your opinion on this?

mkumatag commented:
@ixdy @luxas PTAL above comment let me know your comments please..

luxas commented:
@mkumatag Sounds good to me! Feel free go ahead and implement that!

mkumatag committed:
Multi Arch test images

mkumatag commented:
@luxas modified the code accordingly with the approach mentioned in https://github.com/kubernetes/kubernetes/pull/44910#issuecomment-302306143, PTAL

luxas commented on test/images/image-util.sh:
```diff
@@ -0,0 +1,103 @@
+
+getBaseImage() {
+  arch=$1
+  if [ -f BASEIMAGE ]; then
```
Can't we just check in beforehand that this file exist and just omit the BASEIMAGE sed step if not instead of checking that every time?

luxas commented on test/images/image-util.sh:
```diff
@@ -0,0 +1,103 @@
+getBaseImage() {
+  arch=$1
+  if [ -f BASEIMAGE ]; then
+    for image in $(cat BASEIMAGE)
```
I think this can be done with one line:
```
echo $(grep "${arch}=" BASEIMAGE | cut -d= -f2)
```

luxas commented on test/images/image-util.sh:
```diff
@@ -0,0 +1,103 @@
+    sed -i.bak "s|BASEIMAGE|${BASEIMAGE}|g" Dockerfile
+
+    if grep -q "CROSS_BUILD_" Dockerfile
+    then
```
if [[ foo ]]; then

you have access to bash. Use that pattern consequently in if statements (in other places as well)

luxas commented on test/images/image-util.sh:
```diff
@@ -0,0 +1,103 @@
+push() {
+  archs=$(listArchs)
+  for arch1 in ${archs}; do
+    pushd ${TEST}
```
no need to pushd/popd, you can just use `IMAGENAME=$(${TEST}/NAME)` I think

luxas commented on test/images/iperf/VERSION:
```diff
@@ -0,0 +1 @@
+e2e
```
should we move to a numeric version?

mkumatag replied:
make sense,

luxas commented on test/images/jessie-dnsutils/VERSION:
```diff
@@ -0,0 +1 @@
+e2e
```
numeric version?

luxas commented on test/images/logs-generator/Dockerfile:
```diff
@@ -12,11 +12,11 @@
-CMD /logs-generator --logtostderr \
-  --log-lines-total=${LOGS_GENERATOR_LINES_TOTAL} \
-  --run-duration=${LOGS_GENERATOR_DURATION}
+CMD ["sh", "-c", "/logs-generator --logtostderr --log-lines-total=${LOGS_GENERATOR_LINES_TOTAL} --run-duration=${LOGS_GENERATOR_DURATION}"]
```
Why the extra shell?
`CMD ["/logs-generator", "--logtostderr", "..."]`

mkumatag replied:
This is because of - https://github.com/moby/moby/issues/5509#issuecomment-42173047, anyways I'll revert back to original format without rectangle bracket.

luxas replied:
Ok, leave as-is then :+1:

luxas commented on test/images/mount-tester/Dockerfile:
```diff
@@ -13,5 +13,5 @@
 FROM scratch
-ADD mt mt
-ENTRYPOINT ["/mt"]
+ADD mounttest mounttest
```
COPY mounttest /

luxas commented on test/images/n-way-http/Dockerfile:
```diff
@@ -12,6 +12,6 @@
-ADD server server
-ENTRYPOINT ["/server"]
+FROM BASEIMAGE
+ADD n-way-http n-way-http
```
COPY n-way-http /

luxas commented on test/images/n-way-http/VERSION:
```diff
@@ -0,0 +1 @@
+0.1
```
bump to 0.2 to be sure?

mkumatag replied:
The version in the gcr is completely different and I see version is 1.0. Let me put 1.1 here.

luxas commented on test/images/network-tester/Dockerfile:
```diff
@@ -13,6 +13,6 @@
 
 FROM scratch
-ADD webserver webserver
+ADD nettest nettest
```
COPY nettest /

luxas commented on test/images/network-tester/BASEIMAGE:
```diff
@@ -0,0 +1,5 @@
+amd64=scratch
```
Where all base images are the same, we don't need this file at all

luxas commented on test/images/port-forward-tester/BASEIMAGE:
```diff
@@ -0,0 +1,5 @@
+amd64=scratch
```
unnecessary file

luxas commented on test/images/resource-consumer/.gitignore:
```diff
@@ -0,0 +1,3 @@
+consumer
+consume-cpu/consume-cpu
+controller/controller
```
newline

luxas commented on test/images/resource-consumer/VERSION:
```diff
@@ -0,0 +1 @@
+beta5
```
numeric version please

luxas commented on test/images/resource-consumer/controller/VERSION:
```diff
@@ -0,0 +1 @@
+beta5
```
numeric version please

luxas commented on test/images/serve_hostname/VERSION:
```diff
@@ -0,0 +1 @@
+v1.6
```
v1.8

mkumatag replied:
I checked the gcr and see already multiarch image with different kind of versioning and see v1.6 as the highest version. To avoid any of these confusion let me mention v1.7

luxas commented on test/images/image-util.sh:
```diff
@@ -0,0 +1,103 @@
+    temp_dir=$(mktemp -d)
+    cp -r ${TEST}/* ${temp_dir}
+    if [ -f ${TEST}/bootstrap.sh ]; then
+      pushd ${TEST}
```
Why don't just cd into `${temp_dir}` here, run the script and skip the second arg?
Would be much cleaner

luxas reviewed (commented):
Thanks a lot for your work @mkumatag!

Some general notes:
 - No need to have `BASEIMAGE` when scratch
 - We should now move to numeric versions
 - No need for the second arg to bootstrap.sh
 - You don't have to specify source files to `go build`, but can use `.`
 - Be consistent with your usage of bash and simplify the logic
 - Bump version numbers in yaml in an other PR
 - Is there any case where `IMAGENAME != {directory name}`? Could we make the directory name always match the image name and remove the need for the IMAGENAME file?

mkumatag commented:
> Thanks a lot for your work @mkumatag!
> 
> Some general notes:
> 
> No need to have BASEIMAGE when scratch

The reason for keeping BASEIMAGE to get all the possible archs we are supporting for that particular image. Still we can remove this file assuming that image supported on the architecture
> We should now move to numeric versions

I completely agree with you, we should have numeric version rather than tags like `e2e` or some git sha
> No need for the second arg to bootstrap.sh

I'm creating temporary directory for each arch and copying all the files over to build the dockerimage but go code I'm still compiling from the `test/images/*` and copying over the binary into target temp folder with the help of second param. The reason behind this approach is - I was failed to build few go programs from temp target folder we have deps from `k8s.io/kubernetes/vendor/*` packages which failed to locate while building from temp folder.
> You don't have to specify source files to go build, but can use .

Sure I'll remove wherever I can
> Be consistent with your usage of bash and simplify the logic

Sure
> Bump version numbers in yaml in an other PR

Sure, I'll submit another PR after this.
> Is there any case where IMAGENAME != {directory name}? Could we make the directory name always match the image name and remove the need for the IMAGENAME file?

There are few cases where I see IMAGENAME is not matching the directory names like resource-consumer, volume-tester(multiple dockerimages present)

mkumatag commented on test/images/clusterapi-tester/bootstrap.sh:
```diff
@@ -0,0 +1,7 @@
+SRCS=clusterapi-tester.go
+ARCH=$1
+TARGET=$2
+CGO_ENABLED=0 GOARM=${GOARM} GOARCH=${ARCH} go build -a -installsuffix cgo --ldflags '-w' -o ${TARGET}/${BIN} ./${SRCS}
```
I need source file only for few images where I need to build multiple binaries in the same folder.

mkumatag commented:
@luxas fixed the review comments, PTAL at latest commit.

luxas commented on test/images/dnsutils/BASEIMAGE:
```diff
@@ -0,0 +1,4 @@
+amd64=alpine
```
let's use :3.5 consequently for all alpine instances in this PR

mkumatag replied:
We don't have 3.5 for ppc64le, will keep 3.5 for others except ppc64le

luxas replied:
cc @tianon would it be possible to get 3.6 images for armhf and aarch64 as well?

tianon replied:
https://hub.docker.com/r/arm32v6/alpine/ and https://hub.docker.com/r/arm64v8/alpine/

luxas replied:
@tianon ok. Thank you very much for responding. So armhf is deprecated in favor for arm32v6?
Is there any arm32v7?

tianon replied:
https://github.com/docker-library/official-images#architectures-other-than-amd64 :wink:

luxas replied:
Thanks again @tianon ;)

@mkumatag please implement that in this PR: armhf => arm32v7, aarch64 => arm64v8

luxas commented on test/images/image-util.sh:
```diff
@@ -0,0 +1,108 @@
+
+# Returns list of all supported architectures from BASEIMAGE file
+listArchs() {
+  (
```
why the extra shell?

mkumatag replied:
will remove

luxas commented on test/images/image-util.sh:
```diff
@@ -0,0 +1,108 @@
+
+    docker build --pull -t ${REGISTRY}/${IMAGENAME}-${arch}:${TAG} .
+
+    # Image without any arch postfix will point to amd64 by default.
```
@ixdy as we do this migration, should we keep backwards compat at all?
The reason for keeping `binary:version` for kube-apiserver and such user-facing components is huge, but not for test images like these where there's no consistency at all any way...

I think we can move to `-${ARCH}` variants totally while waiting for gcr.io to support manifest lists, when we can totally skip hardcoding that (yay that day!)

luxas replied:
Still TBD; I'm voting for:
 - moving away from google_containers
 - using a new project with no non-arch versions until we have manifest lists
 - having the directory name be the image name
 - standardizing on format for image tags; e.g. `1.x`, bump x whenever you update

@ixdy Does that sound good to you?

ixdy replied:
* `gcr.io/kubernetes-e2e-test-images` exists now
* I'm in favor of pushing only the `-$arch` versions and removing the legacy amd64 workaround.
* directory name = image name seems like a good idea. it doesn't necessarily need to happen in this PR.
* agree that probably standardizing tags is a good idea. I don't have a strong opinion on exactly what format to use.

luxas commented on test/images/resource-consumer/controller/Dockerfile:
```diff
@@ -12,7 +12,7 @@
 # limitations under the License.
 
-FROM busybox
+FROM scratch
```
Please still use busybox here; add BASEIMAGE

luxas replied:
ping @mkumatag

luxas commented on test/images/resource-consumer/controller/bootstrap.sh:
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/bash -x
```
Why use `-x` everywhere?

mkumatag replied:
I used it for debugging while testing, will remove it.

luxas commented on test/images/resource-consumer/controller/bootstrap.sh:
```diff
@@ -0,0 +1,6 @@
+ARCH=$1
+TARGET=$2
+
+CGO_ENABLED=0 GOARM=${GOARM} GOARCH=${ARCH} go build -a -installsuffix cgo --ldflags '-w' -o ${TARGET}/controller ./controller.go
```
Maybe . can be used here?

mkumatag replied:
sure

luxas commented on test/images/image-util.sh:
```diff
@@ -0,0 +1,108 @@
+    temp_dir=$(mktemp -d)
+    cp -r ${TEST}/* ${temp_dir}
+    if [[ -f ${TEST}/bootstrap.sh ]]; then
+      pushd ${TEST}
```
I'd like to see a little bit more comments in this function yet...

luxas reviewed (commented):
Thanks @mkumatag for your hard work.
Mostly LGTM, but found some nits yet.

Always very happy to cut down more than 300 lines from our codebase :+1: 

As part of this we should align the test images as much as possible; make a standard convention how to write one.

mkumatag commented:
@luxas addressed second round review comments

luxas commented:
@k8s-bot ok to test

@ixdy PTAL, I went through this and commented on most things, some comment might be outstanding, but please give your feedback

mkumatag commented:
@k8s-bot pull-kubernetes-unit test this

ixdy commented on test/images/Makefile:
```diff
@@ -0,0 +1,51 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+REGISTRY ?= gcr.io/google_containers
```
I'd rather not add any more images to google_containers, especially not test images.

maybe we could use gcr.io/k8s-testimages? right now it's largely test-infra images, but it feels like a better fit. what do we think, @kubernetes/sig-testing-misc ?

luxas replied:
@ixdy So this isn't really adding "new" image per se, but just keeping backwards compat with everything that is there now.

Personally I'd be delighted if we could move everything in this dir to something but `k8s-testimages` as well, maybe `k8s-e2e-images` or something. As you said `k8s-testimages` is already being used for test infra stuff...

mkumatag commented:
@ixdy any other comments?

ixdy commented on test/images/dnsutils/VERSION:
```diff
@@ -0,0 +1 @@
+0.1
```
hm.
```
$ gcloud container images list-tags gcr.io/google-containers/dnsutils
DIGEST        TAGS        TIMESTAMP
cd9182f6d74e  e2e         2016-11-20T00:25:38
5c52dce38785  1.0-alpine  2016-09-01T09:05:47
2ee65c65e065              2016-02-19T20:04:41
5c815213864f  latest      2016-02-19T18:35:26
```

ixdy commented on test/images/hostexec/BASEIMAGE:
```diff
@@ -0,0 +1,4 @@
+amd64=alpine:3.6
+arm=armhf/alpine:3.5
```
inconsistency again?

ixdy commented on test/images/iperf/VERSION:
```diff
@@ -0,0 +1 @@
+0.1
```
```
$ gcloud container images list-tags gcr.io/google-containers/iperf
DIGEST        TAGS        TIMESTAMP
ff52c3aa27aa  e2e,latest  2017-01-12T18:54:50
c02c7135b269              2016-05-25T22:35:26
3a3c558d9c62              2016-05-25T22:35:19
e1bc7a99a4d2              2016-04-14T11:45:31
cfc4359451f4              2016-04-14T11:45:26
```

yeah ok this is as good as we can get here

ixdy commented on test/images/jessie-dnsutils/VERSION:
```diff
@@ -0,0 +1 @@
+0.1
```
fyi
```
$ gcloud container images list-tags gcr.io/google-containers/jessie-dnsutils
DIGEST        TAGS    TIMESTAMP
2460d5969122  e2e     2016-02-19T20:09:31
0334c0364283  latest  2016-02-19T18:34:58
```

ixdy commented on test/images/mount-tester-user/BASEIMAGE:
```diff
@@ -0,0 +1,5 @@
+amd64=gcr.io/google_containers/mounttest-amd64:0.9
```
is the mounttest image built first, if we build everything?

ixdy commented on test/images/resource-consumer/BASEIMAGE:
```diff
@@ -0,0 +1,4 @@
+amd64=gcr.io/google-containers/ubuntu-slim-amd64:0.10
```
:12

ixdy commented on test/images/resource-consumer/controller/BASEIMAGE:
```diff
@@ -0,0 +1,4 @@
+amd64=gcr.io/google-containers/ubuntu-slim-amd64:0.10
```
:12

ixdy reviewed (commented):
sorry for the delay, and thanks for taking on this effort!

overall this looks pretty good. my apologies for not following much of the earlier discussion, so I'm not super familiar on some of the design choices.

ixdy commented:
one other note: many of the images use an "e2e" tag right now instead of a numbered version. I'm not sure whether that's a pattern we want to keep or not.

mkumatag commented on test/images/image-util.sh:
```diff
@@ -0,0 +1,115 @@
+      pushd ${TEST}
+      # bootstrap is the program which will take care of all the prerequisites needed
+      # for building the docker image
+      ./bootstrap.sh ${arch} ${temp_dir}
```
As I already explained before to luxas, The reason behind this approach is - I was failed to build few go programs from temp target folder we have deps from k8s.io/kubernetes/vendor/* packages which failed to locate while building from temp folder.

Even this reason is not allowing me to build code inside golang container.. serve_hostname is a very simple go code where they dont depend on any k8s code.!

ixdy replied:
ah, sorry for missing that explanation. dependencies on k8s.io/kubernetes/vendor are a good (and frustrating) point, and probably isn't fixable unless we pull these images out of the kubernetes repo entirely. (which we probably *should* do, but not in this PR. :)

when attempting to build from the temp folder, did you set `GOPATH`?

I'm a little confused about `serve_hostname`, though, since that currently builds inside a docker container without issue.

mkumatag replied:
even GOPATH will not help if you are dependencies are in vendor folder. Only code inside current gopath can use vendor code, so that is the reason I'm compiling code from current gopath and copying over the built binary to target temp folder for building image.

mkumatag commented on test/images/image-util.sh:
```diff
@@ -0,0 +1,115 @@
+        sed -i "s|ARCH|${QEMUARCHS[$arch]}|g" Dockerfile
+        # Register qemu-*-static for all supported processors except the current one
+        docker run --rm --privileged multiarch/qemu-user-static:register --reset
+        curl -sSL https://github.com/multiarch/qemu-user-static/releases/download/${QEMUVERSION}/x86_64_qemu-${QEMUARCHS[$arch]}-static.tar.gz | tar -xz -C ${temp_dir}
```
tar is very small, around ~2MB. I think it should be okay.

mkumatag commented on test/images/Makefile:
```diff
@@ -0,0 +1,51 @@
+sub-push-%:
+	$(MAKE) push-image TEST=$*
+
+all-container: $(addprefix sub-container-,$(WHAT))
```
Initially I had a logic to build all the images in this folder, based on review comment from @luxas I changed logic to make WHAT as compulsory argument,

more infor - https://github.com/kubernetes/kubernetes/pull/44910#discussion_r121389369

mkumatag commented:
@ixdy I've addressed most of your comment, PTAL at my latest commit

ixdy commented on test/images/Makefile:
```diff
@@ -12,19 +12,29 @@
 # limitations under the License.
 
-.PHONY: all image push clean
+REGISTRY ?= k8s-e2e-images
```
maybe on gcr.io, so `gcr.io/k8s-e2e-images`?
guess I should actually create that.
fyi @kubernetes/sig-testing-misc

ixdy commented on test/images/clusterapi-tester/Makefile:
```diff
@@ -12,20 +12,15 @@
-clean:
-	rm -f main
+bin:
+	#CGO_ENABLED=0 GOARM=$(GOARM) GOARCH=$(ARCH) go build -a -installsuffix cgo --ldflags '-w' -o $(TARGET)/$(BIN) .
```
maybe remove this?

ixdy commented on test/images/clusterapi-tester/Makefile:
```diff
@@ -12,20 +12,15 @@
-
-clean:
-	rm -f main
+bin:
```
we probably should also have `.PHONY: bin` in each of these, though it probably won't matter in practice.

ixdy reviewed (approved):
This is awesome work, thanks so much!

 A few minor comments; otherwise, please squash and I think this is good to go.

mkumatag commented:
@ixdy addressed those minor comments and squashed the changes. PTAL and gimme lgtm :)

kargakis commented:
@detiber

luxas commented on test/images/Makefile:
```diff
@@ -12,19 +12,29 @@
 # limitations under the License.
 
-.PHONY: all image push clean
+REGISTRY ?= gcr.io/k8s-e2e-images
```
@ixdy Should we go all in here?
What about the tags then -- should everything start from 1.0 or should they have their cryptic tags as-is now?

ixdy replied:
per some of the discussion around the ci images repo, I finally settled on `gcr.io/kubernetes-e2e-test-images`, so we should update to that.

I agree that our version numbering is pretty inconsistent. (the addons are just as bad.) I guess we can start over, for consistency in the `kubernetes-e2e-test-images` registry, but I don't have a particularly strong opinion on which style to use (e.g. `v1, v2, v3, ...` or `0.1, 0.2, 0.3, ...` or `1.0, 1.1, 1.2, ...` or `v1.0.0` or `$(git rev-parse HEAD)` or `blue, red, green, ...` or ...).
Semantic versioning probably means less for e2e test images, though.

luxas replied:
> gcr.io/kubernetes-e2e-test-images

Sounds perfect!

> 1.0, 1.1, 1.2, ...

I think this is the best one.

luxas commented on test/images/porter/Makefile:
```diff
@@ -1,4 +1,4 @@
-# Copyright 2016 The Kubernetes Authors.
+# Copyright 2017 The Kubernetes Authors.
```
These copyright tags should never be updated; only added when a new file is added

mkumatag replied:
This happened because initially I removed the Makefiles and then created bootstrap and moved them back to Makefile, so resulted in copyright with new year.. :'( I'll take care this..

luxas commented on test/images/serve_hostname/BASEIMAGE:
```diff
@@ -0,0 +1,5 @@
+amd64=busybox
+arm=armhf/busybox
```
all busybox versions should now also be `arm32v6` and `arm64v8`...

mkumatag replied:
sure

luxas commented on test/images/serve_hostname/VERSION:
```diff
@@ -0,0 +1 @@
+v1.7
```
why a `v` prefix here?

mkumatag replied:
This is because the existing version is in this format - https://github.com/kubernetes/kubernetes/blob/master/test/images/serve_hostname/Makefile#L22,

luxas commented on test/images/clusterapi-tester/pod.yaml:
```diff
@@ -4,7 +4,7 @@ metadata:
 spec:
   containers:
-  - image: gcr.io/google_containers/clusterapi-tester:1.0
+  - image: gcr.io/google_containers/clusterapi-tester-amd64:1.1
```
Don't update anything in this PR, that's for the next one :)

mkumatag replied:
oops.. will remove this.

luxas reviewed (commented):
General feedback -- don't make Makefiles executable, don't update old copyright headers, use golang version from toplevel file, be consistent with arm names. 

Also, let's see if the sub-Makefiles can consist of even less code. I'd be fine with renaming the directories with the actual image name. What does @ixdy think of that?

Then we have to decide whether to go all-in on a new repo and re-push all images (my opinion) or keep the old repo.

mkumatag commented on test/images/clusterapi-tester/Makefile:
```diff
@@ -12,20 +12,16 @@
-TAG = 1.0
-PREFIX = gcr.io/google_containers/clusterapi-tester
+bin:
+	docker run -it -v $(TARGET):$(TARGET):Z -v `pwd`/../../../:/go/src/k8s.io/kubernetes:Z \
```
In the past I faced some issue while building with volume mount on selinux enabled systems. This will help me overcome with that issue..

luxas replied:
ok :+1:

mkumatag commented on test/images/fakegitserver/Makefile:
```diff
@@ -12,21 +12,17 @@
-
-clean:
-	rm -f gitserver GITHASH.txt
+.PHONY: bin
```
This I've added based on review comment by @ixdy https://github.com/kubernetes/kubernetes/pull/44910#discussion_r123680354

mkumatag commented on test/images/clusterapi-tester/Makefile:
```diff
@@ -12,20 +12,16 @@
+BIN=clusterapi-tester
+ARCH ?= amd64
+TARGET ?= $(CURDIR)
+GOLANG_VERSION ?= latest
```
if somebody wants to build binary directly using `make bin` this will be used. While calling from image folder this will overridden anyways!

luxas replied:
ok, fine

mkumatag commented:
@luxas fixed your comments.
> General feedback -- don't make Makefiles executable, don't update old copyright headers, use golang version from toplevel file, be consistent with arm names.

- Removed the executable permission for all the makefiles
- Reverted back the copyright year
- I'm already using the GOLANG_VERSION from parent, the image level makefile golang version is to handle just a fallback option if user directly calls that makefile
> 
> Also, let's see if the sub-Makefiles can consist of even less code. I'd be fine with renaming the directories with the actual image name. What does @ixdy think of that?
> 
> Then we have to decide whether to go all-in on a new repo and re-push all images (my opinion) or keep the old repo.

luxas commented on test/images/image-util.sh:
```diff
@@ -0,0 +1,114 @@
+      pushd ${IMAGE}
+      # make bin will take care of all the prerequisites needed
+      # for building the docker image
+      make bin ARCH=${arch} TARGET=${temp_dir}
```
Can you please use `make -C ${IMAGE} bin ...` here instead?

mkumatag replied:
will do it..

luxas commented on test/images/iperf/BASEIMAGE:
```diff
@@ -0,0 +1,4 @@
+amd64=gcr.io/google_containers/ubuntu-slim:0.12
```
@ixdy Would `debian-base` be a better choice than `ubuntu-slim` in the cases where both work?

mkumatag replied:
can we deal optimization later in a separate PR?

ixdy replied:
`debian-base` is probably preferred, yes, though it matters less for test images where security isn't the primary concern.

fine to follow-up in a later PR.

mkumatag replied:
@luxas @ixdy I can work on this to change to debian-base image, but can someone tell are the extra benefit I'll get by doing that?

luxas commented on test/images/image-util.sh:
```diff
@@ -0,0 +1,114 @@
+
+    if [[ -f BASEIMAGE ]]; then
+      BASEIMAGE=$(getBaseImage ${arch})
+      sed -i.bak "s|BASEIMAGE|${BASEIMAGE}|g" Dockerfile
```
any specific reason to use `-i.bak`?

mkumatag replied:
no specif reason, I can remove it.

luxas replied:
ok, please do

luxas commented on test/images/image-util.sh:
```diff
@@ -0,0 +1,114 @@
+      if [[ "${arch}" == "amd64" ]]; then
+        sed -i "/CROSS_BUILD_/d" Dockerfile
+      else
+        sed -i "s|ARCH|${QEMUARCHS[$arch]}|g" Dockerfile
```
Can we use QEMUARCH instead of ARCH here?

luxas commented on test/images/iperf/Dockerfile:
```diff
@@ -12,7 +12,10 @@
-FROM gcr.io/google_containers/ubuntu-slim:0.6
+FROM BASEIMAGE
+
+CROSS_BUILD_COPY qemu-ARCH-static /usr/bin/
```
QEMUARCH instead of ARCH makes it clearer and possible to use ARCH for other purposes...

luxas commented on test/images/jessie-dnsutils/BASEIMAGE:
```diff
@@ -0,0 +1,4 @@
+amd64=debian:jessie
```
Know this was the case before, but wonder if `debian-base` would work here @ixdy ?

mkumatag replied:
can we deal optimization later in a separate PR?

luxas commented on test/images/resource-consumer/Dockerfile:
```diff
@@ -12,7 +12,13 @@
 # limitations under the License.
 
-FROM jess/stress
+FROM BASEIMAGE
```
Would it be possible to set `gcr.io/google-containers/ubuntu-slim-GOARCH:0.12` here and replace GOARCH dynamically?

mkumatag replied:
In the absence of `BASEIMAGE` I'm considering building image for all supported archs like - `amd64, arm, arm64, ppc64le, s390x` and this suits only images based on scratch where we dependent only on the binary image which are packaging.

In the above case I don't want to build this image for `s390x` platform because I'm facing issues while running cross platform using qemu.

So IMO we should have some flexibility to disable any image for any given architecture in the future.

ixdy replied:
@luxas the pattern I've seen in the repo seems to be using `BASEIMAGE` in the Dockerfile, regardless of whether there's a simple pattern for the base image (as is the case with our custom base images). I'm fine with continuing this pattern.

luxas replied:
Ok, fine by me, thanks!

luxas commented on test/images/image-util.sh:
```diff
@@ -0,0 +1,114 @@
+      BASEIMAGE=$(getBaseImage ${arch})
+      sed -i.bak "s|BASEIMAGE|${BASEIMAGE}|g" Dockerfile
+    fi
+
```
could you always replace `GOARCH` with `${arch}` here?

luxas reviewed (commented):
Gets better and better, we have a couple of decisions to make yet...

kargakis commented:
Is this fixing https://github.com/kubernetes/kubernetes/issues/31331?

luxas commented:
Yes, this PR is fixing that issue

> On 24 Jun 2017, at 23:29, Michail Kargakis <notifications@github.com> wrote:
> 
> Is this fixing #31331?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

mkumatag committed:
Port some more images

mkumatag committed:
Move go build to image-utils

mkumatag commented:
@luxas optimized further moving go build code to image-build and calling it in the Makefile, can you PTAL.

mkumatag commented:
@luxas can we consider following things as next phase once we get agreement from all the stakeholders.?

> Still TBD; I'm voting for:
> 
> - moving away from google_containers
> - using a new project with no non-arch versions until we have manifest lists
> - having the directory name be the image name
> - standardizing on format for image tags; e.g. 1.x, bump x whenever you update
> @ixdy Does that sound good to you?

luxas commented:
ping @ixdy PTAL

luxas reviewed (commented):
re-echoing @ixdy's comment here to avoid it being hidden (if that happens)

> gcr.io/kubernetes-e2e-test-images exists now
I'm in favor of pushing only the -$arch versions and removing the legacy amd64 workaround.
directory name = image name seems like a good idea. it doesn't necessarily need to happen in this PR.
agree that probably standardizing tags is a good idea. I don't have a strong opinion on exactly what format to use.

So, what's left to do?
 - Switch to `gcr.io/kubernetes-e2e-test-images`
 - Don't push the legacy `binary` == `binary-amd64`
 - Directory name must match image name => no need for the `NAME` file
 - Set all tags to `1.0`, we're starting from scratch and standardizing all images now.

In a future PR we should document conventions and format for this dir very clearly in the a couple of md files

mkumatag committed:
Fix review comments - luxas, ixdy

mkumatag commented:
@luxas @ixdy fixed the review comments, PTAL

luxas commented:
@mkumatag the verify job wants you to do this:
```
W0630 06:42:56.505] Some packages in hack/.linted_packages do not exist anymore. Please remove them
W0630 06:42:56.505] from hack/.linted_packages:
W0630 06:42:56.505] 
W0630 06:42:56.506]   test/images/mount-tester
W0630 06:42:56.506]   test/images/serve_hostname
W0630 06:42:56.506] 
W0630 06:42:56.506] Some packages passed golint but are not listed in hack/.linted_packages.
W0630 06:42:56.506] Please add them in alphabetical order:
W0630 06:42:56.506] 
W0630 06:42:56.506]   echo test/images/mounttest >> hack/.linted_packages
W0630 06:42:56.506]   echo test/images/serve-hostname >> hack/.linted_packages
W0630 06:42:56.506]   LC_ALL=C sort -o hack/.linted_packages hack/.linted_packages
```

mkumatag committed:
Fix verify-golint

mkumatag commented:
Thanks @luxas.. Fixed that one!

mkumatag commented:
@luxas @ixdy all the test are passed now. Will squash the commits once I get a green signal.

luxas reviewed (approved):
LGTM, thanks @mkumatag!

/lgtm

luxas commented:
Adding the approved label as @ixdy was fine with this and owns those areas as well

mkumatag commented:
Thanks @luxas @ixdy for helping.. Can someone help me push these images?

ixdy commented:
now running: 
```
kubernetes/test/images $ for i in $(find * -maxdepth 0 -type d); do make all-push WHAT=$i; done
```

ixdy commented:
completed. there were some failures (mostly around apt-get signatures failing) that I haven't triaged yet.

ixdy commented:
I can probably look into the missing images during my normal working hours on Friday.

mkumatag commented:
I'm not able to access them via link - https://console.cloud.google.com/gcr/images/kubernetes-e2e-test-images/GLOBAL and complains with error - `You don't have permission to view images in this project.` Not really sure whether this repo is exposed to everyone or not!

mkumatag commented:
> kubernetes/test/images $ for i in $(find * -maxdepth 0 -type d); do make all-push WHAT=$i; done

This command may not push all the image because few images we have multiple folders e.g -  `resource-consumer` and `resource-consumer/controller` 

I prefer listing all the directories which has filename by `VERSION`. VERSION is the common file we have across all the folders for all the test images.

ixdy commented:
Oops, permissions fixed, and default permissions should be fixed for new images, too. PTAL.

Pushing `resource-consumer/controller` now. Good tip regarding looking for the `VERSION` files.
